Userland/Utilities/mknod.cpp

```diff
@@ -1,59 +1,80 @@
-
 ErrorOr<int> serenity_main(Main::Arguments arguments)
 {
     TRY(Core::System::pledge("stdio dpath"));
 
-    // FIXME: Add some kind of option for specifying the file permissions.
-    if (arguments.strings.size() < 3)
-        return usage();
+    StringView mode_string { "0666"sv };
+    StringView name;
+    StringView type_string;
+    StringView major_string;
+    StringView minor_string;
 
-    if (arguments.strings[2].starts_with('p')) {
-        if (arguments.strings.size() != 3)
-            return usage();
-    } else if (arguments.strings.size() != 5) {
-        return usage();
-    }
+    Core::ArgsParser args_parser;
+    args_parser.set_general_help("Creates a file system node.");
+    args_parser.add_option(mode_string, "File mode in octal or symbolic notation", "mode", 'm', "mode");
+    args_parser.add_positional_argument(name, "Pathname to create", "name", Core::ArgsParser::Required::Yes);
+    args_parser.add_positional_argument(
+        type_string,
+        "Type of file to create <b|c|u|p>"
+        "\n\t\tb\tcreate a block special file"
+        "\n\t\tc, u\tcreate a character special file"
+        "\n\t\tp\tcreate a FIFO",
+        "type",
+        Core::ArgsParser::Required::Yes);
+    args_parser.add_positional_argument(major_string, "Major device number", "major", Core::ArgsParser::Required::No);
+    args_parser.add_positional_argument(minor_string, "Minor device number", "minor", Core::ArgsParser::Required::No);
+    args_parser.parse(arguments);
+
+    auto mask = TRY(Core::FilePermissionsMask::parse(mode_string));
+    char type = type_string[0];
 
-    auto name = arguments.strings[1];
-    mode_t mode = 0666;
-    switch (arguments.strings[2][0]) {
+    mode_t mode;
+    switch (type) {
     case 'c':
     case 'u':
-        mode |= S_IFCHR;
+        mode = mask.apply(S_IFCHR);
         break;
     case 'b':
-        mode |= S_IFBLK;
+        mode = mask.apply(S_IFBLK);
         break;
     case 'p':
-        mode |= S_IFIFO;
+        mode = mask.apply(S_IFIFO);
         break;
     default:
-        return usage();
+        warnln("Invalid device type {}", type);
+        return 1;
     }
 
-    int major = 0;
-    int minor = 0;
-    if (arguments.strings.size() == 5) {
-        major = atoi(arguments.argv[3]);
-        minor = atoi(arguments.argv[4]);
+    int major = major_string.to_int().value_or(0);
+    int minor = minor_string.to_int().value_or(0);
+    bool has_none = !major_string.to_int().has_value() && !minor_string.to_int().has_value();
+    bool has_value = major_string.to_int().has_value() && minor_string.to_int().has_value();
+
+    if (type == 'p') {
+        if (!has_none) {
+            warnln("Do not set device numbers when creating FIFO");
+            return 1;
+        }
+    } else {
+        if (!has_value) {
+            warnln("Major and minor device numbers are required");
+            return 1;
+        }
     }
 
     TRY(Core::System::mknod(name, mode, makedev(major, minor)));
```

Comment:
Very minor nitpick: I think you can make the code a bit more concise by rearranging it slightly. Something like:
```suggestion
    auto maybe_major = major_string.to_int();
    auto maybe_minor = minor_string.to_int();
    dev_t device;
    if (type == 'p') {
        if (maybe_major.has_value() || maybe_minor.has_value()) {
            warnln("Do not set device numbers when creating FIFO");
            return 1;
        }
        device = makedev(0, 0);
    } else {
        if (!(maybe_major.has_value() && maybe_minor.has_value())) {
            warnln("Major and minor device numbers are required");
            return 1;
        }
        device = makedev(maybe_major.value(), maybe_minor.value());
    }

    TRY(Core::System::mknod(name, mode, device));
```

Reply:
Good suggestion, thanks